File: 0915. Partition Array into Disjoint Intervals/Solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int partitionDisjoint(vector<int>& nums)
    {
        int n = nums.size(), res = 1, last_max = nums[0], cur_max = nums[0];
        for (int i = 1; i < n; i++) {
            if (nums[i] < last_max) {
                res = i + 1;
                last_max = cur_max;
            } else
                cur_max = max(cur_max, nums[i]);
        }
        return res;
    }
};

File: 0915. Partition Array into Disjoint Intervals/Solution.java
class Solution {
    public int partitionDisjoint(int[] nums) {
        int res = 1, lastMax = nums[0], curMax = nums[0];
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] < lastMax) {
                res = i + 1;
                lastMax = curMax;
            } else {
                curMax = Math.max(curMax, nums[i]);
            }
        }
        return res;
    }
}
